lambdas/package/index.py

```diff
@@ -0,0 +1,210 @@
+    Exception indicating an incomplete response
+    (e.g., from S3 Select)
+    """
+
+
+def load_df(s3response):
+    """
+    Read a streaming response from s3 select into a
+    Pandas DataFrame
+    """
+    buffer = io.StringIO()
+    end_event_received = False
+    stats = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            buffer.write(records)
+        elif 'Progress' in event:
+            print(event['Progress']['Details'])
+        elif 'Stats' in event:
+            stats = event['Stats']['Details']
+        elif 'End' in event:
+            # End event indicates that the request finished successfully
+            end_event_received = True
+
+    if not end_event_received:
+        raise IncompleteResultException("Error: Received an incomplete response from S3 Select.")
+
+    buffer.seek(0)
+    df = pd.read_json(buffer, lines=True)  # pylint: disable=invalid-name
+    return df, stats
+
+
+def load_manifest_detail(s3response):
+    """
+    Read a streaming response from s3 select into a dict
+    """
+    end_event_received = False
+    response_data = None
+    for event in s3response['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode()
+            response_data = json.loads(records)
+        elif 'Progress' in event:
+            print(event['Progress']['Details'])
+        elif 'Stats' in event:
+            print(event['Stats']['Details'])
+        elif 'End' in event:
+            # End event indicates that the request finished successfully
+            end_event_received = True
+
+    if not end_event_received:
+        raise IncompleteResultException("Error: Received an incomplete response from S3 Select.")
+
+    return response_data
+
+
+def get_logical_key_folder_view(df):  # pylint: disable=invalid-name
+    """
+    Post process a set of logical keys to return only the
+    top-level folder view (a special case of the s3-select
+    lambda).
+    """
+    # matches all strings; everything before and including the first
+    # / is extracted
+    folder = df.logical_key.dropna().str.extract('([^/]+/?).*')[0].unique().tolist()
+    return folder
+
+
+def get_s3_client(aws_access_key_id, aws_secret_access_key):
+    """
+    Create an S3 Client using the provided credentials
+    """
+    session = boto3.Session(
+        aws_access_key_id=aws_access_key_id,
+        aws_secret_access_key=aws_secret_access_key
+    )
+    s3_client = session.client('s3')
+    return s3_client
+
+
+def call_s3_select(s3_client, bucket, key, prefix):
+    """
+    Call S3 Select to read only the logical keys from a
+    package manifest that match the desired folder path
+    prefix
+    """
+    sql_stmt = f"SELECT SUBSTRING(s.logical_key, {len(prefix) + 1}) AS logical_key FROM s3object s"
+    if prefix:
+        sql_stmt += f" WHERE SUBSTRING(s.logical_key, 1, {len(prefix)}) = '{prefix}'"
```

Comment:
```suggestion
        quote_str = lambda s: "'%s'" % prefix.replace("'", "''")
        sql_stmt += f" WHERE SUBSTRING(s.logical_key, 1, {len(prefix)}) = {quote_str(prefix)}"
```

Reply:
You're right. We could have single quotes in logical keys. Most SQL libraries I've used have parameter quoting built-in to the API, but I haven't found one for S3 Select. What other characters could we have in logical keys that would break these queries?

Reply:
> Most SQL libraries I've used have parameter quoting built-in to the API, but I haven't found one for S3 Select.

Me too :disappointed: .

> What other characters could we have in logical keys that would break these queries?

This API call fails for chars in range `\x01-\x1f`, except '\t' and '\n', but we might want to forbid them for local keys, since I expect them to not work in other S3 API calls using XML. All other chars seem to work.